README.md
# webrtc live video

## 功能
* web视频直播
* webrtc、ffmpeg、nodejs

## 使用方式

安装ffmpeg [安装脚本]

安装node-webrtc [跳转]
```
npm install
npm start
带摄像头的手机或本子，使用新版chrome或firefox打开：localhost:3000，实时视频推流到node server
使用其他设备打开：localhost:3000/live
```

[安装脚本]: ./tool/install.sh
[跳转]: https://js-platform.github.io/node-webrtc/

HISTORY.md
# Changelog
All notable changes to this project will be documented in this file.

## [0.1.0] - 2018-06-28
### Changed
- readme完善

## 0.0.1 - 2018-03-03
### Added



peer.js
const webrtc = require('wrtc')
const WebSocket = require('ws')
const { Readable }  = require('stream')
const ffmpeg = require('fluent-ffmpeg')
const path = require('path')

const ws = new WebSocket('ws://localhost:3000')

const pc = new webrtc.RTCPeerConnection({iceServers: [{urls: "stun:stun.services.mozilla.com"}]})
const MAX_BUFFERED_AMOUNT = 64 * 1024
const rs = new Readable()
rs._read = () => {}

let channel
let channelName
let channelReady = false

const begin = Date.now()
ffmpeg()
.input(rs)
.videoCodec('libx264')
.audioCodec('copy')
.addOption('-threads', 2)
.addOption('-f', 'hls')
.addOption('-hls_segment_filename', 'static/video-dist2/segment%03d.ts')
.addOption('-hls_time', 5)
.addOption('-hls_list_size', 6)
.addOption('-preset', 'veryfast')
.on('end', function () {
  console.log('file has been converted succesfully,time:', Date.now() - begin);
})
.on('error', function (err) {
  console.log('an error happened: ' + err.message);
})
.save(path.join(__dirname, './static/video-dist2/video.m3u8'))


// pc.onaddstream = function (obj) {
//   debugger
//   console.log(obj)
// }

pc.ondatachannel = function(event) {
  if (!event.channel) {
    throw(new Error('Data channel event is missing `channel` property'))
  }
  channel = event.channel
  channel.binaryType = 'arraybuffer'

  if (typeof channel.bufferedAmountLowThreshold === 'number') {
    channel.bufferedAmountLowThreshold = MAX_BUFFERED_AMOUNT
  }
  channelName = channel.label

  channel.onmessage = function (event) {
    if (!channelReady) {
      return
    }
    let data = event.data
    if (data instanceof ArrayBuffer) {
      data = Buffer.from(data)
    }
    rs.push(data)
  }
  channel.onbufferedamountlow = function () {
    throw(new Error('buffered amount low'))
  }
  channel.onopen = function () {
    channelReady = true
  }
  channel.onclose = function () {
    console.log('close')
  }
  channel.onerror = function (err) {
    throw(err)
  }
}

pc.onicecandidate = function(e) {
  if (e.candidate) {
    ws.send(JSON.stringify({ type: 'candidate', candidate: e.candidate }))
  }
}
const errorCallback = (err) => {
  console.log(err)
  pc.close()
}

new Promise(resolve => ws.addEventListener('open', resolve))
.then(() => {
  ws.send(JSON.stringify({ type: 'ok' }))
  return new Promise(
    resolve => ws.addEventListener('message', (event) => {
      const data = JSON.parse(event.data)
      if (data.type === 'ok') {
        resolve()
      }
      if (data.type === 'offer') {
        pc.setRemoteDescription(new webrtc.RTCSessionDescription(data.offer), function() {
          pc.createAnswer(function(answer) {
            pc.setLocalDescription(new webrtc.RTCSessionDescription(answer), function() {
              ws.send(JSON.stringify({ type: 'answer', answer }))
            }, errorCallback)
          }, errorCallback)
        }, errorCallback)
      }
      if (data.type === 'candidate') {
        pc.addIceCandidate(new webrtc.RTCIceCandidate(data.candidate))
      }
    })
  )
})
.catch(errorCallback)


src/chat.js
import polyfill from './polyfill'

const { RTCPeerConnection, RTCSessionDescription, playMediaStream, getUserMedia } = polyfill
const $localVideo = window.document.getElementById('local-video')

const playVideo = ($video, stream) => playMediaStream($video, stream)

class Peer {
  constructor() {
    this.isCaller = !window.location.search
    this.ws = new WebSocket('ws://' + window.location.host)
    this.pc = new RTCPeerConnection({iceServers: [{urls: "stun:stun.services.mozilla.com"}]})
    this.channel = this.pc.createDataChannel("Mydata")
    this.init()
  }
  init() {
    const ws = this.ws
    const pc = this.pc
    const channel = this.channel

    const errorCallback = this.errorCallback.bind(this)
    new Promise(resolve => ws.addEventListener('open', resolve))
    .then(() => { window.onbeforeunload = () => ws.close() })
    .then(() => {
      // pc.onaddstream = obj => playVideo($remoteVideo, obj.stream)

      pc.onicecandidate = (e) => {
        if (e.candidate) {
          ws.send(JSON.stringify({ type: 'candidate', candidate: e.candidate }))
        }
      }
    })
    .then(() => getUserMedia({
      audio: true,
      video: true
    }))
    .then((mediastream) => {
      playVideo($localVideo, mediastream)
      channel.onopen = () => {
        let top = 0
        let btm = 0
        let buffer = []
        const mediaRecorder = new window.MediaRecorder(mediastream)
        mediaRecorder.ondataavailable = (e) => {
          const reader = new window.FileReader()
          reader.readAsArrayBuffer(e.data)
          const index = top++

          reader.addEventListener('loadend', function () {
            // const buf = Buffer.from(new Uint8Array(reader.result))
            
            const buf = new Uint8Array(reader.result)
            let i = index - btm
            while (buffer.length < i) buffer.push(null)
            buffer[i] = buf
            while (buffer.length && buffer[0]) {
              var next = buffer.shift()
              btm++
              channel.send(next)
            }
          })
        }
        mediaRecorder.start(1000)
        // stream.on('data', chunk => channel.send(chunk))
      }
      // this.pc.addStream(stream)
    })
    .then(() => {
      ws.send(JSON.stringify({ type: 'ok' }))
      return new Promise(
        resolve => ws.addEventListener('message', (event) => {
          const data = JSON.parse(event.data)
          if (data.type === 'ok') {
            resolve()
          }
        })
      )
    })
    .then(() => {
      if (this.isCaller) {
        pc.createOffer(function (offer) {
          pc.setLocalDescription(new RTCSessionDescription(offer), function () {
            ws.send(JSON.stringify({ type: 'offer', offer }))
          }, errorCallback)
        }, errorCallback, {
          offerToReceiveAudio: 1,
          offerToReceiveVideo: 1
        })
        ws.addEventListener('message', function (event) {
          const data = JSON.parse(event.data)
          if (data.type === 'answer') {
            pc.setRemoteDescription(new RTCSessionDescription(data.answer), function() {
            }, errorCallback)
          }
          if (data.type === 'candidate') {
            pc.addIceCandidate(new RTCIceCandidate(data.candidate))
          }
        });
      } else {
        ws.addEventListener('message', function incoming(event) {
          const data = JSON.parse(event.data)
          if (data.type === 'offer') {
            pc.setRemoteDescription(new RTCSessionDescription(data.offer), function() {
              pc.createAnswer(function(answer) {
                pc.setLocalDescription(new RTCSessionDescription(answer), function() {
                  ws.send(JSON.stringify({ type: 'answer', answer }))
                }, errorCallback)
              }, errorCallback)
            }, errorCallback)
          }
          if (data.type === 'candidate') {
            pc.addIceCandidate(new RTCIceCandidate(data.candidate))
          }
        })
      }
    })
    .catch(errorCallback)
  }
  errorCallback(err) {
    console.log(err)
    this.pc.close()
    this.ws.close()
  }
}

new Peer()

src/live.js
import polyfill from './polyfill'

const { RTCPeerConnection, RTCSessionDescription, playMediaStream, getUserMedia } = polyfill
const $localVideo = window.document.getElementById('local-video')
const $remoteVideo = window.document.getElementById('remote-video')

const playVideo = ($video, stream) => playMediaStream($video, stream)

class Peer {
  constructor() {
    this.isCaller = !window.location.search
    this.ws = new WebSocket('ws://' + window.location.host)
    this.pc = new RTCPeerConnection({iceServers: [{urls: "stun:stun.services.mozilla.com"}]})
    this.init()
  }
  init() {
    const ws = this.ws
    const pc = this.pc
    const errorCallback = this.errorCallback.bind(this)
    new Promise(resolve => this.ws.addEventListener('open', resolve))
    .then(() => { window.onbeforeunload = () => ws.close() })
    .then(() => {
      this.pc.onaddstream = obj => playVideo($remoteVideo, obj.stream)
      this.pc.onicecandidate = (e) => {
        if (e.candidate) {
          this.ws.send(JSON.stringify({ type: 'candidate', candidate: e.candidate }))
        }
      }
    })
    .then(() => getUserMedia({
      audio: true,
      video: true
    }))
    .then((stream) => {
      playVideo($localVideo, stream)
      this.pc.addStream(stream)
    })
    .then(() => {
      ws.send(JSON.stringify({ type: 'ok' }))
      return new Promise(
        resolve => ws.addEventListener('message', (event) => {
          const data = JSON.parse(event.data)
          if (data.type === 'ok') {
            resolve()
          }
        })
      )
    })
    .then(() => {
      if (this.isCaller) {
        pc.createOffer(function (offer) {
          pc.setLocalDescription(new RTCSessionDescription(offer), function () {
            ws.send(JSON.stringify({ type: 'offer', offer }))
          }, errorCallback)
        }, errorCallback, {
          offerToReceiveAudio: 1,
          offerToReceiveVideo: 1
        })
        ws.addEventListener('message', function (event) {
          const data = JSON.parse(event.data)
          if (data.type === 'answer') {
            pc.setRemoteDescription(new RTCSessionDescription(data.answer), function() {
            }, errorCallback)
          }
          if (data.type === 'candidate') {
            pc.addIceCandidate(new RTCIceCandidate(data.candidate))
          }
        });
      } else {
        ws.addEventListener('message', function incoming(event) {
          const data = JSON.parse(event.data)
          if (data.type === 'offer') {
            pc.setRemoteDescription(new RTCSessionDescription(data.offer), function() {
              pc.createAnswer(function(answer) {
                pc.setLocalDescription(new RTCSessionDescription(answer), function() {
                  ws.send(JSON.stringify({ type: 'answer', answer }))
                }, errorCallback)
              }, errorCallback)
            }, errorCallback)
          }
          if (data.type === 'candidate') {
            pc.addIceCandidate(new RTCIceCandidate(data.candidate))
          }
        })
      }
    })
    .catch(errorCallback)
  }
  errorCallback(err) {
    console.log(err)
    this.pc.close()
    this.ws.close()
  }
}

new Peer()

src/polyfill.js
const URL = window.URL || window.webkitURL || window.mozURL || window.msURL
const RTCPeerConnection = window.RTCPeerConnection || window.webkitRTCPeerConnection || window.mozRTCPeerConnection
const RTCSessionDescription = window.RTCSessionDescription || window.webkitRTCSessionDescription || window.mozRTCSessionDescription

let getUserMedia = navigator.mediaDevices && navigator.mediaDevices.getUserMedia.bind(navigator.mediaDevices)

if (!getUserMedia) {
  getUserMedia = (constraints) => {
    const temp = navigator.getUserMedia || navigator.webkitGetUserMedia || navigator.mozGetUserMedia || Promise.reject(new Error('getUserMedia is not implemented in this browser'))

    return new Promise((resolve, reject) => temp.call(navigator, constraints, resolve, reject))
  }
}
const playMediaStream = ($video, stream) => {
  const SRC_OBJECT = 'srcObject' in $video ? 'srcObject' : ('mozSrcObject' in $video ? 'mozSrcObject' : ('webkitSrcObject' in $video ? 'webkitSrcObject' : null))
  if (SRC_OBJECT) {
    $video[SRC_OBJECT] = stream
  } else {
    $video.src = URL && URL.createObjectURL(stream) || stream
  }
  $video.onloadedmetadata = function (e) {
    $video.play()
  }
}

export default { RTCPeerConnection, RTCSessionDescription, playMediaStream, getUserMedia }


tool/install.sh
#!/usr/bin/env bash

#下载
wget ftp://ftp.videolan.org/pub/x264/snapshots/last_x264.tar.bz2
wget https://nchc.dl.sourceforge.net/project/lame/lame/3.100/lame-3.100.tar.gz
wget http://ffmpeg.org/releases/ffmpeg-4.0.1.tar.bz2

#解压
tar xvf last_x264.tar.bz2
tar zxvf lame-3.100.tar.gz
tar xvf ffmpeg-4.0.1.tar.bz2

#安装x264
./configure --prefix=/usr/local/x264 --enable-static --bit-depth=10 --disable-asm
make
make install

#安装lame
./configure --prefix=/usr/local/lame --enable-static
make && make install

#安装ffmpeg
export PKG_CONFIG_PATH=/usr/local/x264/lib/pkgconfig:$PKG_CONFIG_PATH

./configure --enable-gpl --enable-libx264 --enable-libmp3lame --enable-shared --prefix=/usr/local/ffmpeg --extra-cflags=-I/usr/local/lame/include --extra-cxxflags=-I/usr/local/lame/include --extra-ldflags=-L/usr/local/lame/lib
#编译时发现ffmpeg的一个问题，会提示x264_bit_depth未定义，需要修改libavcodec/libx264.c源码，变量名改成X264_BIT_DEPTH即可
make
make install

vim /etc/ld.so.conf
#/usr/local/ffmpeg/lib/加入动态链接配置
ldconfig

#测试
ffmpeg -i video.mp4 -c:v libx264 -c:a copy -f hls -threads 8 -hls_time 5 -hls_list_size 0 video.m3u8
ffmpeg -i audio.mp3  -ab 64000 -ar 44100 -acodec libmp3lame ceshi.mp3

server.js
const Koa = require('koa')
const WebSocket = require('ws')
const serve = require('koa-static')
const path = require('path')
const http = require('http')
const peer = require('./peer')
// const koaBody = require('koa-body')

const app = new Koa()
app.use(serve(path.join(__dirname, './static/'), {
  // maxage: 30 * 24 * 3600 * 1000,
  maxage: 0
}))
// app.use(koaBody())
const server = http.createServer(app.callback())
const wss = new WebSocket.Server({ server })

wss.broadcast = function broadcast(data) {
  wss.clients.forEach(function each(client) {
    if (client.readyState === WebSocket.OPEN) {
      client.send(data)
    }
  })
}

const chat = {
  owner: false,
  client: false
}

let count = 0
wss.on('connection', function connection(ws, req) {
  // You might use location.query.access_token to authenticate or share sessions
  // or req.headers.cookie (see http://stackoverflow.com/a/16395220/151312)

  ws.on('message', function incoming(message) {
    const json = JSON.parse(message)
    if (json.type === 'ok') {
      count += 1
      if (count === 2) {
        wss.broadcast(JSON.stringify({ type: 'ok' }))
      }
    }
    // console.log('received: %s', message)
    const others = Array.from(wss.clients).filter(client => client !== ws && client.readyState === WebSocket.OPEN)
    if (others.length > 0) {
      others.forEach(client => client.send(message))
    }
  })
  ws.on('close', () => {
    count -= 1
  })
})

server.listen(3000, function listening() {
  console.log('Listening on %d', server.address().port)
})